pkg/icingadb/history/retention.go

```diff
@@ -173,6 +173,13 @@ func (r *Retention) Start(ctx context.Context) error {
 			continue
 		}
 
+		if time.Now().AddDate(0, 0, -int(days)).Before(time.Unix(0, 0)) {
+			r.logger.Errorf("Skipping history retention for category %s: "+
+				"%d days of retention will result in negative timestamps, overflowing unsigned timestamps",
+				stmt.Category, days)
+			continue
+		}
```

Comment:
My first though: this could be moved closer to where the overflow would happen. This is the function that will actually use the value:

https://github.com/Icinga/icingadb/blob/a6d134c7f3e2d41e9ae2453f08e9672430f9c188/pkg/icingadb/cleanup.go#L21-L30

Adding the following right at the beginning would actually be a valid implementation of the function:

```go
// Timestamps in the database are unsigned, so there can't be any rows with a timestamp smaller than
// a negative value and the query can be skipped.
if olderThan.Before(time.Unix(0, 0)) {
	return 0, nil
}
```

My second though: oh, looks like there's actually no overflow at all. Like the timestamp is only passed around as `UnixMilli` in the code:

https://github.com/Icinga/icingadb/blob/a6d134c7f3e2d41e9ae2453f08e9672430f9c188/pkg/icingadb/cleanup.go#L40-L43

And it doesn't seem to ever be converted to `uint64`. And the `Value()` implementation of `UnixMilli` just returns the `int64` returned by [`time.Time.UnixMilli()`](https://pkg.go.dev/time#Time.UnixMilli):
https://github.com/Icinga/icinga-go-library/blob/bb5b3747011a3d20befc17dafa0757f2eb3eb1e6/types/unix_milli.go#L87-L95

So it will just generate a query with something like `WHERE event_time < -42` which won't match any rows but should be fine otherwise I guess? So I doubt we need this `if` at all. Like running in an overflow in scenario sounds plausible, but not with the current implementation.

Reply:
> My first though: this could be moved closer to where the overflow would happen. This is the function that will actually use the value:
> 
> https://github.com/Icinga/icingadb/blob/a6d134c7f3e2d41e9ae2453f08e9672430f9c188/pkg/icingadb/cleanup.go#L21-L30
> 
> Adding the following right at the beginning would actually be a valid implementation of the function:
> 
> ```go
> // Timestamps in the database are unsigned, so there can't be any rows with a timestamp smaller than
> // a negative value and the query can be skipped.
> if olderThan.Before(time.Unix(0, 0)) {
> 	return 0, nil
> }
> ```

A similar idea also came to me, but if the condition is true at the current position, it will also be true (at least for a while) within the `periodic.Start` callback function. Thus, the function will be a no-op within a loop. Doesn't result in load, but seems a bit unnecessary.

> My second though: oh, looks like there's actually no overflow at all. Like the timestamp is only passed around as `UnixMilli` in the code:
> 
> https://github.com/Icinga/icingadb/blob/a6d134c7f3e2d41e9ae2453f08e9672430f9c188/pkg/icingadb/cleanup.go#L40-L43
> 
> And it doesn't seem to ever be converted to `uint64`. And the `Value()` implementation of `UnixMilli` just returns the `int64` returned by [`time.Time.UnixMilli()`](https://pkg.go.dev/time#Time.UnixMilli): https://github.com/Icinga/icinga-go-library/blob/bb5b3747011a3d20befc17dafa0757f2eb3eb1e6/types/unix_milli.go#L87-L95
> 
> So it will just generate a query with something like `WHERE event_time < -42` which won't match any rows but should be fine otherwise I guess? So I doubt we need this `if` at all. Like running in an overflow in scenario sounds plausible, but not with the current implementation.

This seems right. However, I wanted to verify this with setting `retention.history-days` to 60,000 - ending up somewhere in the 1860s. After enabling debug logs, I found this query:
```
26 Execute  DELETE FROM notification_history WHERE environment_id = '\\Nha\\\/\\\\\\)W\\\' AND send_time < -3455367174796 ORDER BY send_time LIMIT 5000
```

The timestamp seems valid:
```
date -d @-3455367174.796
Tue Jul  3 08:40:33 AM LMT 1860
```

Manually running queries with negative timestamps also works, of couse. Thus, yes, you are right. I have removed the check now :)